### PR TITLE
[Blazor] Endpoint discovery

### PR DESCRIPTION
**Pull request by javiercn**
https://github.com/dotnet/aspnetcore/issues/46984
https://github.com/dotnet/aspnetcore/issues/46985

I've implemented the "scaffold" for discovery even though we don't do discovery. The goal of the implementation is to avoid reflection completely and support source generation and allowing the user to tweak the results via endpoint conventions.

We are not doing discovery beyond the current assembly, and only for pages. We are also not allowing the user to customize the list of assemblies or pages that we discover just yet.

The approach that we are trying to follow is simpler than what MVC does, but we might end up with a similar design when we implement discovery entirely.

We rely on having a component that acts as the root/entry point of the application to discover the rest of the pages that are part of the application. Our entry point implements an interface (which we will automatically source generate a definition and implementation for based on the calls to `MapRazorComponents`) that returns a `ComponentApplicationBuilder`.

This builder is used internally to provide the list of pages along with their routes and other metadata. The builder is passed down to the `RazorComponentConventionEndpointBuilder`, which will be the piece that exposes additional methods to add or remove entire assemblies from the mappings or individual sets of types.

We use the data in the builder when we build the list of endpoints to generate the initial endpoint definition, which the user can the further customize via conventions.

* We expect customization to happen via conventions.
  * We don't think we need an application model like MVC, as endpoints are already customizable.
* We expect any reflection to happen in the default implementations for the `ComponentApplicationBuilder`.
  * The goal here is to source generate the implementation and avoid reflection completely elsewhere in the system.

### CODE REVIEW

**Comment by davidfowl** [comment]
We should be looking at a source generator here...

**Comment by javiercn** [comment]
@davidfowl that's the plan, but its more complicated than that, because the pages are generated on the Razor compiler, which is also a source generator, so we need to either do it on the Razor Compiler or have a feature on the Compiler to allow some level of ordering on source generators.

**Comment by davidfowl** [comment]
Feels like the right place to do it is in the razor compiler 😄. Dropping a breadcrumb that we can use at runtime or the method itself.

**Comment by javiercn** [comment]
@davidfowl The plan is to have something that is eventually source generated, and for the design to take that into account.

> Dropping a breadcrumb that we can use at runtime or the method itself.

Not sure exactly what you mean. (Beware, that I am still experimenting on the shape this thing will have).

**Comment by davidfowl** [comment]
> @davidfowl The plan is to have something that is eventually source generated, and for the design to take that into account.

I don't think we should do it eventually; I think that should be the primary design.

**Comment by javiercn** [comment]
@davidfowl let me rephrase it.

The final thing is going to be source generated, that's the primary design. We do, however, don't want to make source generation a requirement initially, hence why we have a runtime strategy to fallback, so that we:
1) Make the change in stages.
2) Minimize risk if we run into issues with the SG strategy.

But to be absolutely clear, source generation is the way forward. The main challenges for source generation are that we can't just write one, because it can't interact with the compiler. So that's a risk we need to mitigate.

**Comment by davidfowl** [comment]
We don't do anything like this in the framework and we shouldn't start now. Even if we don't get the full source generation support now, we should not be scanning the entire reference assembly graph looking for components. Not even MVC is doing this (which was already a mistake) and we shouldn't be adding new code like this to the framework.

File: src/Components/Samples/BlazorServerApp/App.razor [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿<Router AppAssembly="@typeof(Program).Assembly">
+﻿@implements IRazorComponentApplication<App>
```
Comment (SteveSandersonMS) [inline src/Components/Samples/BlazorServerApp/App.razor:1]: Is this required? Assuming it's not (for back-compat) then maybe we could not have this here, and only have it on the United sample.

**Review by SteveSandersonMS — APPROVED** [review]
Looks good. We discussed, and although you mentioned adding more tests, I'll approve now to ensure this is not blocked for preview 3.

Longer term we may want to consider other ways of indicating a root component without needing the `IRazorComponentApplication<T>` interface which is a little unusual. For example we could keep the API as `MapRazorPages<TRootComponent>()` and have the logic be something like:

* Look for assembly attributes on `typeof(TRootComponent).Assembly`
* If we can find one like `[assembly:RazorComponentsApplicationManifest(typeof(SourceGeneratedType))]` then we do `Activator.CreateInstance(thatType)` and call a `GetListOfPages` method on it
* If not, we assume there isn't a source generator yet and fall back on the existing reflection-based logic

Then, no `IRazorComponentApplication<T>` declarations are needed anywhere.

As we discussed that would use a bit more reflection but AFAIK it is the kind that has no drawbacks with AOT and there's only a small, bounded amount of logic running here on startup assuming the source generator did run (regardless of the number of pages) so it should not adversely affect startup times.

File: src/Components/Samples/BlazorUnitedApp/Shared/MainLayout.razor [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ﻿@inherits LayoutComponentBase
+@implements IRazorComponentApplication<MainLayout>
```
Comment (javiercn) [inline src/Components/Samples/BlazorUnitedApp/Shared/MainLayout.razor:2]: Ideally, we are going to source generate this interface definition and implementation based on a convention (like you have an App.razor file on your project) or another gesture (like extending a special type if we find a reason to have one)

File: src/Components/Samples/BlazorUnitedApp/Program.cs [hunk 3]
```diff
@@ -26,6 +27,6 @@
 
 app.UseRouting();
 
-app.MapRazorComponents();
+app.MapRazorComponents<MainLayout>();
```
Comment (javiercn) [inline src/Components/Samples/BlazorUnitedApp/Program.cs:30]: This might be replaced in the future with `app.MapRazorComponents<App>` if we end up having a root component for Blazor united as we do for Webassembly and Server.

File: src/Components/Samples/BlazorServerApp/App.razor (hunk 1 above)
Comment (javiercn) [reply src/Components/Samples/BlazorServerApp/App.razor:1]: Ah, not sure why this ended up here. I'll remove it.

Comment (javiercn) [reply src/Components/Samples/BlazorServerApp/App.razor:1]:
```suggestion
```

**Review by MackinnonBuck — APPROVED** [review]

File: src/Components/Samples/BlazorUnitedApp/Shared/MainLayout.razor (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Samples/BlazorUnitedApp/Shared/MainLayout.razor:2]: If that works out without drawbacks, that could be a good solution.

In many ways it's extremely strange that having a call to `DoSomething<T>` actually *causes* the type `T` to implement some other interface. We'll have to be super cautious about only considering this in cases where we can be completely sure it will never cause some other compiler error because people would be totally mystified about what's going on. And I'm not sure how it will make sense if `T` is not in the current project.

File: src/Components/Samples/BlazorUnitedApp/Program.cs (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/Samples/BlazorUnitedApp/Program.cs:30]: As it stands, the type `T` here is just a way to indicate the containing assembly, right? It's unfortunate that C# does not have an `assemblyof(MyApp)` (where `MyApp` is not a type, but rather is one of the assemblies in the compilation) keyword but I agree this would make sense if `T` goes on to serve the role of a root component.
